File: javascript/resistor-color-trio/resistor-color-trio.js
//
// This is only a SKELETON file for the 'Resistor Color Trio' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const colorValues = [
	"black",
	"brown",
	"red",
	"orange",
	"yellow",
	"green",
	"blue",
	"violet",
	"grey",
	"white",
];

function colorToValue(col) {
	const index = colorValues.indexOf(col);

	if (index === -1) throw new Error(/invalid color/);

	return index;
}

function toUnit(num) {
	if (num >= 1000) return "kiloohms";
	return "ohms";
}

export class ResistorColorTrio {
	constructor([first, second, exp]) {
		this.value = colorToValue(first) * 10 + colorToValue(second);

		this.value *= 10 ** colorToValue(exp);

		this.unit = toUnit(this.value);

		if (this.unit === "kiloohms") this.value /= 1000;
	}

	get label() {
		return `Resistor value: ${this.value} ${this.unit}`;
	}
}

File: python/resistor-color-trio/resistor_color_trio.py
from typing import Literal

COLORS = [
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
]

METRIC_PREFIXES = [
    "",
    "kilo",
    "mega",
    "giga"
]


def label(colors: list[str]) -> str:
    val = value(colors)
    exponent = color_code(colors[2])

    total = val * 10 ** exponent

    zeroes = 0
    while total > 999:
        if total % 1000 == 0:
            zeroes += 3
        total /= 1000
    
    prefix = METRIC_PREFIXES[zeroes // 3]
    return str(int(total)) + " " + prefix + "ohms"


def value(colors: list[str]) -> int:
    total = 0
    for color in colors[:2]:
        total = total * 10 + color_code(color)
    return total


def color_code(color: str) -> int:
    return COLORS.index(color)
